File: benchmark/src/inverse.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.util.Scanner;

public class inverse {
    static void swapRows(double[][] matrix, double[][] inverse, int row1, int row2, int n) {
        for (int i = 0; i < n; i++) {
            double temp = matrix[row1][i];
            matrix[row1][i] = matrix[row2][i];
            matrix[row2][i] = temp;
            temp = inverse[row1][i];
            inverse[row1][i] = inverse[row2][i];
            inverse[row2][i] = temp;
        }
    }

    static void gaussJordan(double[][] matrix, double[][] inverse, int n) {
        for (int i = 0; i < n; i++) {
            if (matrix[i][i] == 0) {
                for (int j = i + 1; j < n; j++) {
                    if (matrix[j][i] != 0) {
                        swapRows(matrix, inverse, i, j, n);
                        break;
                    }
                }
            }
            double pivot = matrix[i][i];
            for (int j = 0; j < n; j++) {
                matrix[i][j] /= pivot;
                inverse[i][j] /= pivot;
            }
            for (int j = 0; j < n; j++) {
                if (j != i) {
                    double factor = matrix[j][i];
                    for (int k = 0; k < n; k++) {
                        matrix[j][k] -= factor * matrix[i][k];
                        inverse[j][k] -= factor * inverse[i][k];
                    }
                }
            }
        }
    }

    public static void main(String[] args) {
        try {
            Scanner inputFile = new Scanner(new File("input/input.txt"));
            PrintWriter outputFile = new PrintWriter("output/java.txt");

            int n = inputFile.nextInt();
            double[][] matrix = new double[n][n];
            double[][] inverse = new double[n][n];

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    matrix[i][j] = inputFile.nextDouble();
                    inverse[i][j] = (i == j) ? 1.0 : 0.0;
                }
            }

            gaussJordan(matrix, inverse, n);

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    outputFile.print(inverse[i][j] + " ");
                }
                outputFile.println();
            }

            inputFile.close();
            outputFile.close();
        } catch (FileNotFoundException e) {
            System.out.println("Error opening file!");
        }
    }
}

File: benchmark/src/inverse.scala
import scala.io.Source
import java.io.PrintWriter

object Inverse {
  def swapRows(matrix: Array[Array[Double]], row1: Int, row2: Int): Unit = {
    val temp = matrix(row1)
    matrix(row1) = matrix(row2)
    matrix(row2) = temp
  }

  def gaussJordan(matrix: Array[Array[Double]], inverse: Array[Array[Double]], n: Int): Unit = {
    for (i <- 0 until n) {
      // If the pivot is zero, swap with a lower row
      if (matrix(i)(i) == 0) {
        for (j <- (i + 1) until n) {
          if (matrix(j)(i) != 0) {
            swapRows(matrix, i, j)
            swapRows(inverse, i, j)
            // Break the inner loop after swapping
            return
          }
        }
      }
      val pivot = matrix(i)(i)
      for (j <- 0 until n) {
        matrix(i)(j) /= pivot
        inverse(i)(j) /= pivot
      }
      for (j <- 0 until n) {
        if (j != i) {
          val factor = matrix(j)(i)
          for (k <- 0 until n) {
            matrix(j)(k) -= factor * matrix(i)(k)
            inverse(j)(k) -= factor * inverse(i)(k)
          }
        }
      }
    }
  }

  def main(args: Array[String]): Unit = {
    val inputFile = "input/input.txt"
    val outputFile = "output/scala.txt"
    val lines = Source.fromFile(inputFile).getLines().toArray
    val n = lines(0).toInt

    val matrix = Array.ofDim[Double](n, n)
    val inverse = Array.ofDim[Double](n, n)

    for (i <- 0 until n) {
      val rowValues = lines(i + 1).split(" ").map(_.toDouble)
      for (j <- 0 until n) {
        matrix(i)(j) = rowValues(j)
        inverse(i)(j) = if (i == j) 1.0 else 0.0
      }
    }

    gaussJordan(matrix, inverse, n)

    val writer = new PrintWriter(outputFile)
    for (i <- 0 until n) {
      writer.println(inverse(i).mkString(" "))
    }
    writer.close()
  }
}
